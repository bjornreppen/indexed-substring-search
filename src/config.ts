export const config = {
  tokenize: (phrase: string) => {
    if (phrase.indexOf(" ") < 0) {
      return [phrase];
    }
    return phrase.split(/[.,\/\:\(\) -]/);
  }
};
