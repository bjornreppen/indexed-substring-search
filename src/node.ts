export interface INode {
  [key: string]: any;
}

export interface IResult {
  key: any;
  score: number;
}
