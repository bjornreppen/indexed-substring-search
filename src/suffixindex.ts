import { config } from "./config";
import { INode } from "./node";

export class SuffixIndex {
  private index: INode;

  constructor(index: INode) {
    this.index = index;
  }

  /** Look up one or more substrings, return the set matching both */
  public queryPhrase(q: string) {
    const words = config.tokenize(q);
    const wordMatches: any = [];
    words.forEach(word => {
      wordMatches.push(this.queryWord(word));
    });
    // For our intersect, sort the shortest result set first
    wordMatches.sort((a: any, b: any) => a.length - b.length);
    const result: any = [];
    const [first, ...rest] = wordMatches;
    Object.keys(first).forEach((key: string) => {
      let score = first[key];
      for (const rx of rest) {
        if (!(key in rx)) {
          return;
        }
        score *= rx[key];
      }
      result.push({ key: key, score: score });
    });
    result.sort((a: any, b: any) => b.score - a.score);
    return result;
  }

  /** Look up a single word in the index */
  public queryWord(q: string) {
    q = q.toLowerCase();
    let current = this.index;
    for (const char of q) {
      if (!current[char]) {
        return {};
      }
      current = current[char];
    }
    const results = this.reduce(current, {});
    return results;
    //      return results.sort((a: any, b: any) => {
    //      return (Object as any).values(b)[0] - (Object as any).values(a)[0];
    //  });
  }

  private reduce(r: INode, s: any, parentKey: string = "") {
    Object.keys(r).forEach(key => {
      if (parentKey === "$") {
        s[key] = Math.max(s[key] || 0, r[key]);
      } else {
        this.reduce(r[key], s, key);
      }
    });
    return s;
  }
}
