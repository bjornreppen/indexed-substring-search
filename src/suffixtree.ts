import * as fs from "fs";
import { config } from "./config";
import { INode, IResult } from "./node";
import { SuffixIndex } from "./suffixindex";

/** A tiny substring search module
 * https://en.wikipedia.org/wiki/Suffix_tree
 * Can be optimized massively especially for space, but trading size for developer hours.
 */
export class SuffixTree {
  private index: INode;

  constructor(json = {}) {
    this.index = json;
  }

  /** Serialize the tree to string */
  public toJson() {
    return JSON.stringify(this.index);
  }

  /** Builds the qeryable index. */
  public buildIndex() {
    return new SuffixIndex(this.index);
  }

  /** index a sequence of words */
  public addSentence(sentence: string, searchResult: IResult) {
    if (!sentence) {
      return;
    }
    const words = config.tokenize(sentence);
    for (const word of words) {
      this.addWord(word, searchResult);
    }
  }

  /** index a single word */
  private addWord(word: string, searchResult: IResult) {
    if (word.length < 2) {
      return;
    }
    for (let i = 0; i < word.length; i++) {
      this.addSuffix(word.substring(i).toLowerCase(), searchResult);
    }
  }

  /** index ending of a string */
  private addSuffix(sub: string, searchResult: IResult) {
    let current = this.index;
    for (const char of sub) {
      if (!current[char]) {
        current[char] = {};
      }
      current = current[char];
    }
    if (!current.$) {
      current.$ = {};
    }
    current.$[searchResult.key] = searchResult.score;
  }
}
