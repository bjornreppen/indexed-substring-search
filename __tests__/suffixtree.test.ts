import { SuffixTree } from "../src/";

test("Banana", () => {
  const tree = new SuffixTree();
  tree.addSentence("banana", { key: "BANANA", score: 1 });
  expect(tree).toMatchSnapshot();
});

test("Banana search nana", () => {
  const tree = new SuffixTree();
  tree.addSentence("banana", { key: "BANANA", score: 1 });
  const index = tree.buildIndex();
  expect(index.queryWord("nana")).toMatchSnapshot();
});

test("Banana ban search 'nana ban'", () => {
  const tree = new SuffixTree();
  tree.addSentence("banana ban", { key: "BANANA", score: 1 });
  const index = tree.buildIndex();
  expect(index.queryPhrase("nana ban")).toMatchSnapshot();
});

test("Multiple hits sharing phrases", () => {
  const actual = query(
    [
      { phrase: "wood", key: "wood", score: 1 },
      { phrase: "branch", key: "wood", score: 1 }
    ],
    "wood"
  );
  expect(actual).toMatchSnapshot();
});

test("i myra eller skogen", () => {
  const actual = query(
    [
      { phrase: "in the woods", key: "myr", score: 1 },
      { phrase: "in the forest", key: "skog", score: 0.9 }
    ],
    "the in"
  );
  expect(actual).toMatchSnapshot();
});

const query = (indexInput: any[], query: string) => {
  const tree = new SuffixTree();
  indexInput.forEach(e =>
    tree.addSentence(e.phrase, { key: e.key, score: e.score })
  );
  const index = tree.buildIndex();
  return index.queryPhrase(query);
};
