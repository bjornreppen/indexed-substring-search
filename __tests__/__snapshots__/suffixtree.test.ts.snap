// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banana 1`] = `
SuffixTree {
  "index": Object {
    "a": Object {
      "$": Object {
        "BANANA": 1,
      },
      "n": Object {
        "a": Object {
          "$": Object {
            "BANANA": 1,
          },
          "n": Object {
            "a": Object {
              "$": Object {
                "BANANA": 1,
              },
            },
          },
        },
      },
    },
    "b": Object {
      "a": Object {
        "n": Object {
          "a": Object {
            "n": Object {
              "a": Object {
                "$": Object {
                  "BANANA": 1,
                },
              },
            },
          },
        },
      },
    },
    "n": Object {
      "a": Object {
        "$": Object {
          "BANANA": 1,
        },
        "n": Object {
          "a": Object {
            "$": Object {
              "BANANA": 1,
            },
          },
        },
      },
    },
  },
}
`;

exports[`Banana ban search 'nana ban' 1`] = `
Array [
  Object {
    "key": "BANANA",
    "score": 1,
  },
]
`;

exports[`Banana search nana 1`] = `
Object {
  "BANANA": 1,
}
`;

exports[`Multiple hits sharing phrases 1`] = `
Array [
  Object {
    "key": "wood",
    "score": 1,
  },
]
`;

exports[`i myra eller skogen 1`] = `
Array [
  Object {
    "key": "myr",
    "score": 1,
  },
  Object {
    "key": "skog",
    "score": 0.81,
  },
]
`;
